### ISSUE TITLE
Retry cannot be cancelled

### ISSUE THREAD

**Issue by martin-sladecek** [oracle/helidon#3650]
There is no way to cancel Retry FT handler.
 
## Environment Details
* Helidon Version: 2.4.0
* Helidon SE or Helidon MP
* JDK version: 15
* OS: Linux
* Docker version (if applicable):

----------

## Problem Description
The `Single` instance that comes from `io.helidon.faulttolerance.Retry` `invoke` method has a cancel() method. Calling this method should stop Retry handler, but it just cancels the internal CompletableFuture and keeps on with the retries.

## Steps to reproduce
```

    public static ScheduledExecutorService EXECUTOR_SERVICE = Executors.newScheduledThreadPool(2);

    public static void main(String[] args) throws Exception{
        Single<Void> invoke = Retry.builder()
                .retryPolicy(Retry.JitterRetryPolicy.builder().calls(100).delay(Duration.ofSeconds(1))
                        .jitter(Duration.ZERO).build())
                .overallTimeout(Duration.ofSeconds(2000)).build().invoke(
                        () -> CompletableFuture.runAsync(() -> {
                            try {
                                System.out.println("Executing task");
                                Thread.sleep(1000);
                            } catch (Throwable e) {
                                e.printStackTrace();
                            }
                            throw new RuntimeException("Retry!");
                        }, EXECUTOR_SERVICE));
        invoke.exceptionallyAccept(Throwable::printStackTrace);
;
        Thread.sleep(2000);
        System.out.println("Calling cancel()");
        invoke.cancel();
    }
```

**Comment by martin-sladecek** [comment]
Also, shouldn't all FtHandler implementations actually cancel the provided CompletionStage (if it is a CompletableFuture)?

This is what SingleFromCompletionStage is currently doing for the "source" CompletionStage, however, none of the FtHandler implementations actually call it.

There are at least 2 groups:
1. FtHandlers that use DelayedTask, which actually uses its internal "resultFuture" CompletableFuture as a source for the SingleFromCompletionStage, which means the original CompletionStage will not be cancelled. 
2. TimeoutImpl is actually using the provided CompletionStage as a source
```
    @Override
    public <T> Single<T> invoke(Supplier<? extends CompletionStage<T>> supplier) {
        if (!currentThread) {
            return Single.create(supplier.get(), true)
                    .timeout(timeoutMillis, TimeUnit.MILLISECONDS, executor.get());
```
so I would expect cancel() to be called on that stage in this case, but it is not.

So using `FtHandler.invoke(() -> completionStage)` works differently than simple `Single.create(completionStage)`, including 
`Timeout.invoke(..)` vs `Single.create(...).timeout(...)` - the former does not cancel the supplied stage while the latter does.

I've just stating my observations here as I'm not sure whether I understand the intended behaviour of the FtHandler.cancel(). The current behaviour is just confusing to me.

**Comment by spericas** [comment]
@martin-sladecek Let me start reproducible test, just to make sure I understand the intent here:

1. I'm assuming the overall timeout is just to make it bigger than 1 sec, and as such is not really part of the problem here. If so, I'll set it to a much larger number for my testing.

2. The example runs a task for 1 second, waits a second (delay) and then runs the first retry. From a quick run, the first retry is actually interrupted, but happens to print the message before entering the sleep. If there's another print after the try-catch, it is not shown. OTOH, I don't see the stack trace printed due to the interruption (hmmm).

**Comment by spericas** [comment]
@martin-sladecek I recall implementing `SingleFromCompletionStage::cancel` a while back for Fault Tolerance MP. I'll take a look, it may indeed be that we are not propagating calls to `cancel` the way we should in SE.

**Comment by martin-sladecek** [comment]
> * I'm assuming the overall timeout is just to make it bigger than 1 sec, and as such is not really part of the problem here. If so, I'll set it to a much larger number for my testing.
Right, overall timeout is there just so that other sleeps can be bigger.

> * The example runs a task for 1 second, waits a second (delay) and then runs the first retry. From a quick run, the first retry is actually interrupted, but happens to print the message before entering the sleep. If there's another print after the try-catch, it is not shown. OTOH, I don't see the stack trace printed due to the interruption (hmmm).
cancel() does *not* interrupt the processing, this is consistent with the CompletableFuture behaviour and is fine. What (I think) I saw was rather that after cancel, a second retry started immediately and I had tasks running in parallel, because the first one cannot be interrupted through CompletableFuture.

> Is this the sort of use case that you have in mind?
Yes, exactly. SingleFromCompletionStage does this, but it seems like this code is not executed for FT handlers for some reason.

**Comment by spericas** [comment]
PR #3680 
PR #3682
